ai8x.py

```diff
@@ -563,6 +572,148 @@ def get_activation(activation=None):
     return Empty()
 
 
+def histogram(inp, bins):
+    """
+    CUDA compatible histogram calculation
+    """
+    minimum, maximum = inp.min(), inp.max()
+    counts = torch.histc(inp, bins, min=minimum, max=maximum).cpu()
+    boundaries = torch.linspace(minimum, maximum, bins + 1)
+    return counts, boundaries
+
+
+def calc_q_error(module, threshold, bits, eps=1e-9):
+    """
+    Activation quantization error calculation
+    """
+    quantized_hist = module.hist[1].clone()
+    quantized_hist = torch.round((quantized_hist / (threshold + eps)) * 2**(bits-1))
+    quantized_hist = torch.clamp(quantized_hist, -2**(bits-1), 2**(bits-1)-1)
+    quantized_hist = (quantized_hist * (threshold + eps) / 2**(bits-1))
+    err = torch.sum(((quantized_hist - module.hist[1])**2)*module.hist[0]) \
+        / torch.sum(module.hist[0])
+
+    return err
+
+
+def _merge_hist(module):
+    """
+    Merge histograms of activations
+    """
+    bins_to_stack = []
+    for hist in module.hist:
+        bins_to_stack.append(hist[1])
+    stacked_bins = torch.stack(bins_to_stack)
+    min_edge = stacked_bins.min()
+    max_edge = stacked_bins.max()
+    # 2048 is the number of bins
+    width = (max_edge - min_edge) / 2048
+    merged_bins = torch.arange(min_edge.item(), (max_edge+width).item(), width.item())
+    merged_counts = None
+
+    for hist in module.hist:
+        if merged_counts is None:
+            merged_counts = _interpolate_hist(hist[0], hist[1], merged_bins)
+        else:
+            merged_counts += _interpolate_hist(hist[0], hist[1], merged_bins)
+
+    module.hist = (merged_counts, merged_bins)
+
+
+def _interpolate_hist(counts, bins, new_bins):
+    """
+    Helper function for interpolating histograms to new bins
+    """
+    cumulative_hist = torch.cumsum(counts, dim=0).to(device=bins.device)
+    cumulative_hist = torch.cat((torch.tensor([0]), cumulative_hist))
+    cumulative_interp_hist = torch.from_numpy(np.interp(new_bins.numpy(), bins.numpy(),
+                                                        cumulative_hist.numpy()))
+    interp_counts = torch.diff(cumulative_interp_hist, prepend=torch.tensor([0]))
+
+    return interp_counts
```

Comment:
I'm not a fan of the additional computations of calculating cumulative sums than taking the differences at the end.

I would suggest `interp_counts = torch.from_numpy(np.interp(new_bins.numpy(), bins.numpy(), np.concatenate(([0],counts.numpy()))))*(new_bins.max()-new_bins.min())/(bins.max()-bins.min())`